third_party/amd/lib/TritonAMDGPUToLLVM/Utility.cpp

```diff
@@ -1,17 +1,223 @@
 #include "Utility.h"
+#include "ConvertLayoutOpToLLVM/SharedToDotOperandHelper.h"
 #include "PatternTritonGPUOpToLLVM.h"
+#include "TargetInfo.h"
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
 #include "mlir/Dialect/LLVMIR/LLVMTypes.h"
 #include "mlir/Dialect/LLVMIR/NVVMDialect.h"
 #include "mlir/Dialect/LLVMIR/ROCDLDialect.h"
+#include "mlir/IR/BuiltinTypes.h"
 #include "mlir/IR/PatternMatch.h"
+#include "mlir/IR/TypeRange.h"
+#include "mlir/IR/TypeUtilities.h"
+#include "mlir/Support/LLVM.h"
+#include "triton/Analysis/Utility.h"
 #include "triton/Conversion/TritonGPUToLLVM/TypeConverter.h"
 #include "triton/Conversion/TritonGPUToLLVM/Utility.h"
 #include "triton/Dialect/Triton/IR/Dialect.h"
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/TypeSwitch.h"
+#include "llvm/Support/LogicalResult.h"
+#include <utility>
 
 using mlir::triton::gpu::appendOrGetExternFuncOp;
 using mlir::triton::gpu::getFunctionType;
 
 namespace {
+
+// Utility function to traverse a struct, get to the GEP contained
+// in the struct at position `pos` and extract its base pointer and offset
+mlir::FailureOr<std::pair<Value, Value>> getBaseAndOffset(Value ptr,
+                                                          int64_t pos = 0) {
+  // Typedef the return type here, to have more coincise code
+  using ReturnType = mlir::FailureOr<std::pair<Value, Value>>;
+
+  Operation *currentOp = ptr.getDefiningOp();
+  auto res =
+      llvm::TypeSwitch<Operation *, ReturnType>(currentOp)
+          .Case<LLVM::GEPOp>([&](auto gepOp) -> ReturnType {
+            SmallVector<Value> indices =
+                llvm::to_vector(gepOp.getDynamicIndices());
+            if (indices.size() == 1)
+              return std::make_pair(gepOp.getBase(), indices[0]);
+            return failure();
+          })
+          .Case<LLVM::AddrSpaceCastOp>([&](auto addrspaceCastOp) -> ReturnType {
+            return getBaseAndOffset(addrspaceCastOp.getArg(), pos);
+          })
+          .Case<LLVM::ExtractValueOp>([&](auto extractValOp) -> ReturnType {
+            ArrayRef<int64_t> position = extractValOp.getPosition();
+            if (position.size() > 1)
+              return failure();
+            return getBaseAndOffset(extractValOp.getContainer(), position[0]);
+          })
+          .Case<LLVM::InsertValueOp>([&](auto insertValOp) -> ReturnType {
+            ArrayRef<int64_t> position = insertValOp.getPosition();
+            if (position.size() > 1)
+              return failure();
+            if (position[0] == pos)
+              return getBaseAndOffset(insertValOp.getValue(), 0);
+            return getBaseAndOffset(insertValOp.getContainer(), pos);
+          })
+          .Default([&](Operation *op) -> ReturnType { return failure(); });
+  return res;
+}
+
+// Utility class to take care of buffer operation emission. We may add more
+// emitters into this as needed.
```

Comment:
Can we put some comment here to explain in high level how the oob is handled? Basically we set numRecords to `i32_max`, and use an offset `i32_max + 1`.